x-pack/plugins/visualization_lens/public/types.ts

```diff
@@ -0,0 +1,176 @@
+  //     setState: (newState: T) => void
+  //   ) => DatasourcePublicAPI[key]
+  // };
+}
+
+/**
+ * This is an API provided to visualizations by the frame, which calls the publicAPI on the datasource
+ */
+export interface DatasourcePublicAPI {
+  getTableSpec: () => TableSpec;
+  getOperationForColumnId: () => Operation;
+
+  // Called by dimension
+  getDimensionPanelComponent: (
+    props: DatasourceDimensionPanelProps
+  ) => (domElement: Element, operations: Operation[]) => void;
+
+  removeColumnInTableSpec: (columnId: string) => TableSpec;
+  moveColumnTo: (columnId: string, targetIndex: number) => void;
+  duplicateColumn: (columnId: string) => TableSpec;
+}
+
+export interface DatasourceDataPanelProps {
+  domElement: Element;
+}
+
+// The only way a visualization has to restrict the query building
+export interface DatasourceDimensionPanelProps {
+  // If no columnId is passed, it will render as empty
+  columnId?: string;
+
+  // Visualizations can restrict operations based on their own rules
+  filterOperations: (operation: Operation) => boolean;
+
+  // Visualizations can hint at the role this dimension would play, which
+  // affects the ordering of the query
+  suggestedPriority?: DimensionPriority;
+}
+
+export interface DatasourceValidationError {
+  type: 'error';
+}
+
+export interface DatasourceValidationValid {
+  type: 'valid';
+}
+
+export type DataType = 'string' | 'number' | 'date' | 'boolean';
+
+// An operation does not represent a column in the datatable
+export interface Operation {
+  // Operation ID is a reference to the operation
+  id: string;
+  // User-facing label for the operation
+  label: string;
+  dataType: DataType;
+  // A bucketed operation has many values the same
+  isBucketed: boolean;
+
+  // Extra meta-information like cardinality, color
+}
+
+export interface TableSpecColumn {
+  // Column IDs are the keys for internal state in data sources and visualizations
+  columnId: string;
+}
+
+// TableSpec is managed by visualizations
+export type TableSpec = TableSpecColumn[];
+
+export type VisualizationTableRequest = Array<{
+  dataType: 'string' | 'number';
+  isBucketed: boolean;
+}>;
+
+export interface VisualizationProps<T> {
+  datasource: DatasourcePublicAPI;
+  state: T;
+  setState: (newState: T) => void;
+}
+
+export interface VisualizationSuggestion<T> {
+  score: number;
+  title: string;
+  state: T;
+  datasourceSuggestionId: string;
+}
+
+export interface Visualization<T> {
+  // Used to switch to this visualization from another
+  getInitialStateFromOtherVisualization: (
+    options: {
+      roles: DimensionRole[];
+      datasource: DatasourcePublicAPI;
+      state?: T;
+    }
+  ) => T[];
+
+  renderConfigPanel: (props: VisualizationProps<T>) => void;
+
+  toExpression: (state: T, datasource: DatasourcePublicAPI) => string;
+
+  // For use in transitioning from one viz to another
+  getMappingOfTableToRoles: (state: T, datasource: DatasourcePublicAPI) => DimensionRole[];
```

Comment:
Same here. Maybe we punt on this kind of transition until later. Another perk of that is that it keeps our initial workload smaller.